lib/archethic/utils/typed_encoding.ex

```diff
@@ -0,0 +1,147 @@
+defmodule Archethic.Utils.TypedEncoding do
+  @moduledoc """
+  Handle encoding of typed data
+  """
+
+  @type_int 0
+  @type_float 1
+  @type_str 2
+  @type_list 3
+  @type_map 4
+  @type_bool 5
+  @type_nil 6
+
+  @type arg() :: number() | boolean() | binary() | list() | map() | nil
+
+  alias Archethic.Utils
+  alias Archethic.Utils.VarInt
+
+  @spec serialize(arg(), bit_size :: pos_integer()) :: bitstring()
+  def serialize(int, bit_size) when is_integer(int) do
```

Comment:
To my own it would make more sense to have the serialization mode in entry of this function and regarding it the serialization could adapt the bit size.
```elixir
TypedEncoding.serialize(value, :compact)
```

```elixir
serialize(value, :compact) do
    do_serialize(value, 1)
end

serialize(value, :extended) do
    do_serialize(value, 8)
end
```

This would avoid having case on serialization mode before calling the TypedEncoding.serialize function

And I think the responsability to determine what is the bit size of a compact encoding is to the TypedEncoding module, not the responsability of the module which calls it